lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1015 @@
+
+    // Try to find the right module for this interface, either alongside it,
+    // in the cache, or in the prebuilt cache.
+    SmallVector<FileDependency, 16> allDeps;
+    auto moduleOrErr =
+      discoverUpToDateModuleForInterface(modulePath, cachedOutputPath, allDeps);
+
+    // If we errored with anything other than 'no such file or directory',
+    // fail this load and let the other module loader diagnose it.
+    if (!moduleOrErr &&
+        moduleOrErr.getError() != std::errc::no_such_file_or_directory)
+      return moduleOrErr.getError();
+
+    // We discovered a module! Return the module's path so we know what to load.
+    if (moduleOrErr) {
+      auto &module = *moduleOrErr;
+      // If it's prebuilt, use this time to generate a forwarding module.
+      if (module.isPrebuilt())
+        if (writeForwardingModule(module, cachedOutputPath, allDeps))
+          return std::make_error_code(std::errc::not_supported);
+
+      return module.getModulePath();
+    }
+
+    // We didn't discover a module corresponding to this interface. Build one.
+    if (builder.buildSwiftModule(cachedOutputPath, /*shouldSerializeDeps*/true))
+      return std::make_error_code(std::errc::invalid_argument);
+    return cachedOutputPath.str();
+  }
+};
+
+} // end anonymous namespace
+
+/// Load a .swiftmodule associated with a .swiftinterface either from a
+/// cache or by converting it in a subordinate \c CompilerInstance, caching
+/// the results.
+std::error_code ParseableInterfaceModuleLoader::findModuleFilesInDirectory(
+  AccessPathElem ModuleID, StringRef DirPath, StringRef ModuleFilename,
+  StringRef ModuleDocFilename,
+  std::unique_ptr<llvm::MemoryBuffer> *ModuleBuffer,
+  std::unique_ptr<llvm::MemoryBuffer> *ModuleDocBuffer) {
+
+  // If running in OnlySerialized mode, ParseableInterfaceModuleLoader
+  // should not have been constructed at all.
+  assert(LoadMode != ModuleLoadingMode::OnlySerialized);
+
+  auto &fs = *Ctx.SourceMgr.getFileSystem();
+  llvm::SmallString<256> ModPath, InPath;
+
+  // First check to see if the .swiftinterface exists at all. Bail if not.
+  ModPath = DirPath;
+  path::append(ModPath, ModuleFilename);
+
+  auto Ext = file_types::getExtension(file_types::TY_SwiftParseableInterfaceFile);
+  InPath = ModPath;
+  path::replace_extension(InPath, Ext);
+  if (!fs.exists(InPath))
+    return std::make_error_code(std::errc::no_such_file_or_directory);
+
+  // Create an instance of the Impl to do the heavy lifting.
+  ParseableInterfaceModuleLoaderImpl impl(
+                Ctx, ModPath, InPath, ModuleID.first.str(),
+                CacheDir, PrebuiltCacheDir, ModuleID.second, dependencyTracker,
+                LoadMode);
+
+  // Ask the impl to find us a module that we can load or give us an error
+  // telling us that we couldn't load it.
+  auto pathOrErr = impl.findOrBuildLoadableModule();
+  if (!pathOrErr)
+    return pathOrErr.getError();
+  std::string finalPath = *pathOrErr;
```

Comment:
Nitpick: This is going to copy the path. `std::move`?

Also nitpick: it's a little disconcerting to be mixing UpperCamelCase and lowerCamelCase for local variables in the same function. Can you convert this to be all one style or the other afterwards?